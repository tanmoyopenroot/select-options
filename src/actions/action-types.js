/**
 * Action types for redux actions
 * @module actions/action-types
 */

export const FETCH_DATA = 'FETCH_DATA';
export const FETCHED_USERS_DATA = 'FETCHED_USERS_DATA';
export const FETCHED_ROLES_DATA = 'FETCHED_ROLES_DATA';
export const FETCHED_PROJECT_DATA = 'FETCHED_PROJECT_DATA';

export const SELECTED_DATA = 'SELECTED_DATA';

export const SET_SELECTED_USER = 'SET_SELECTED_USER';
export const SET_SELECTED_ROLE = 'SET_SELECTED_ROLE';
export const SET_SELECTED_PROJECT = 'SET_SELECTED_PROJECT';
