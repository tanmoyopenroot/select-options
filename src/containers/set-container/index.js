export { default } from './set-container';
