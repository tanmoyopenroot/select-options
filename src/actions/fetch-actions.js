import { FETCH_DATA } from './action-types';

export const fetchData = () => ({
  type: FETCH_DATA
})
