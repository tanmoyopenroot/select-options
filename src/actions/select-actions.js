import { SELECTED_DATA } from './action-types';

export const saveSelectData = (type, set, id) => ({
  type: SELECTED_DATA,
  payload: {
    type,
    setID: set,
    id,
  }
})
