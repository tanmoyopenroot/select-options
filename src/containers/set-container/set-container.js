import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import shallowCompare from 'react-addons-shallow-compare';

import SelectComponent from '../../components/select-component';
import { saveSelectData } from '../../actions/select-actions';
import { getUsers } from '../../selectors/users-selector';
import { getRoles } from '../../selectors/roles-selector';
import { getProjects } from '../../selectors/projects-selector';

const FlexContainer = styled.div`
  display: flex;
  flex-wrap: nowrap;
  margin: 0 auto;
  width: 80%;
`;

const Header = styled.h2`
  font-size: 20px;
  font-weight: normal;
  width: 80%;
  margin: 30px auto;
`;

export class SetContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  shouldComponentUpdate(nextProps, nextState) {
    return shallowCompare(this, nextProps, nextState);
  }

  render () {
    const { setID, users, roles, projects, saveSelectData } = this.props;
    const { usersID, usersHash } = users;
    const { rolesID, rolesHash } = roles; 
    const { projectsID, projectsHash } = projects; 

    return (
      <Fragment>
        <Header>
          Set Number {setID} 
        </Header>
        <FlexContainer>
          <SelectComponent
            key={`set${setID}-users`}
            IDList={usersID}
            data={usersHash}
            type='users'
            setID={setID}
            saveSelectData={saveSelectData}
          />
          <SelectComponent
            key={`set${setID}-roles`}
            IDList={rolesID}
            data={rolesHash}
            type='roles'
            setID={setID}
            saveSelectData={saveSelectData}
          />
          <SelectComponent
            key={`set${setID}-projects`}
            IDList={projectsID}
            data={projectsHash}
            type='projects'
            setID={setID}
            saveSelectData={saveSelectData}
          />
        </FlexContainer>
      </Fragment>
    );
  }
}

SetContainer.propTypes = {
  setID: PropTypes.number, 
  saveSelectData: PropTypes.func,
  users: PropTypes.object.isRequired,
  roles: PropTypes.object,
  projects: PropTypes.object,
}

const mapDispatchToProps = {
  saveSelectData,
}

const mapStateToProps = state => ({
  users: getUsers(state),
  roles: getRoles(state),
  projects: getProjects(state),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SetContainer);
