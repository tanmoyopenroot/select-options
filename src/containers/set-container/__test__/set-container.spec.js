import React from 'react';
import { SetContainer } from '../set-container';
import { shallow } from 'enzyme';

import { saveSelectData } from '../../../actions/select-actions';

const setup = () => {
  const users = {
    usersID: [1, 2, 3],
    usersHash: {
      1: { id: 1, name: 'John Doe'}, 
      2: { id: 2, name: 'Alice' },
      3: { id: 3, name: 'Bob' }
    },
  }
  
  const roles = {
    rolesID: [1, 2, 3],
    rolesHash: {
      1: { id: 1, name: 'Admin'}, 
      2: { id: 2, name: 'Editor' },
      3: { id: 3, name: 'Viewer' }
    },
  }
  
  const projects = {
    projectsID: [1, 2, 3],
    projectsHash: {
      1: { id: 1, name: 'Trip to space'},
      2: { id: 2, name: 'Assembly Ikea furniture' },
      3: { id: 3, name: 'Datumize Zentral' }
    },
  }

  return {
    users,
    roles,
    projects
  }
} 

describe('Testing <SetContainer />', () => {
  const { users, roles, projects } = setup();
    
  const wrapper = shallow(
    <SetContainer
      key={1}
      setID={1}
      saveSelectData={saveSelectData}
      users={users}
      roles={roles}
      projects={projects}
    />,
  )

  test('Testing SetContainer\'s props', () => { 
    expect(wrapper.length).toEqual(1);
    expect(wrapper.children().length).toBe(2);
    expect(Object.keys(wrapper.instance().props).length).toEqual(5);
    expect(wrapper.instance().props.users).toEqual(users);
    expect(wrapper.instance().props.roles).toEqual(roles);
    expect(wrapper.instance().props.projects).toEqual(projects);
    expect(wrapper).toMatchSnapshot();
  });
})
