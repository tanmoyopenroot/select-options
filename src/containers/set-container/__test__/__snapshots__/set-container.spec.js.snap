// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <SetContainer /> Testing SetContainer's props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SetContainer
    projects={
        Object {
            "projectsHash": Object {
              "1": Object {
                "id": 1,
                "name": "Trip to space",
              },
              "2": Object {
                "id": 2,
                "name": "Assembly Ikea furniture",
              },
              "3": Object {
                "id": 3,
                "name": "Datumize Zentral",
              },
            },
            "projectsID": Array [
              1,
              2,
              3,
            ],
          }
    }
    roles={
        Object {
            "rolesHash": Object {
              "1": Object {
                "id": 1,
                "name": "Admin",
              },
              "2": Object {
                "id": 2,
                "name": "Editor",
              },
              "3": Object {
                "id": 3,
                "name": "Viewer",
              },
            },
            "rolesID": Array [
              1,
              2,
              3,
            ],
          }
    }
    saveSelectData={[Function]}
    setID={1}
    users={
        Object {
            "usersHash": Object {
              "1": Object {
                "id": 1,
                "name": "John Doe",
              },
              "2": Object {
                "id": 2,
                "name": "Alice",
              },
              "3": Object {
                "id": 3,
                "name": "Bob",
              },
            },
            "usersID": Array [
              1,
              2,
              3,
            ],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <styled.h2>
          Set Number 
          1
</styled.h2>,
        <styled.div>
          <SelectComponent
                    IDList={
                              Array [
                                        1,
                                        2,
                                        3,
                                      ]
                    }
                    data={
                              Object {
                                        "1": Object {
                                          "id": 1,
                                          "name": "John Doe",
                                        },
                                        "2": Object {
                                          "id": 2,
                                          "name": "Alice",
                                        },
                                        "3": Object {
                                          "id": 3,
                                          "name": "Bob",
                                        },
                                      }
                    }
                    saveSelectData={[Function]}
                    setID={1}
                    type="users"
          />
          <SelectComponent
                    IDList={
                              Array [
                                        1,
                                        2,
                                        3,
                                      ]
                    }
                    data={
                              Object {
                                        "1": Object {
                                          "id": 1,
                                          "name": "Admin",
                                        },
                                        "2": Object {
                                          "id": 2,
                                          "name": "Editor",
                                        },
                                        "3": Object {
                                          "id": 3,
                                          "name": "Viewer",
                                        },
                                      }
                    }
                    saveSelectData={[Function]}
                    setID={1}
                    type="roles"
          />
          <SelectComponent
                    IDList={
                              Array [
                                        1,
                                        2,
                                        3,
                                      ]
                    }
                    data={
                              Object {
                                        "1": Object {
                                          "id": 1,
                                          "name": "Trip to space",
                                        },
                                        "2": Object {
                                          "id": 2,
                                          "name": "Assembly Ikea furniture",
                                        },
                                        "3": Object {
                                          "id": 3,
                                          "name": "Datumize Zentral",
                                        },
                                      }
                    }
                    saveSelectData={[Function]}
                    setID={1}
                    type="projects"
          />
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "Set Number ",
            1,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Set Number ",
          1,
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <SelectComponent
              IDList={
                            Array [
                                          1,
                                          2,
                                          3,
                                        ]
              }
              data={
                            Object {
                                          "1": Object {
                                            "id": 1,
                                            "name": "John Doe",
                                          },
                                          "2": Object {
                                            "id": 2,
                                            "name": "Alice",
                                          },
                                          "3": Object {
                                            "id": 3,
                                            "name": "Bob",
                                          },
                                        }
              }
              saveSelectData={[Function]}
              setID={1}
              type="users"
/>,
            <SelectComponent
              IDList={
                            Array [
                                          1,
                                          2,
                                          3,
                                        ]
              }
              data={
                            Object {
                                          "1": Object {
                                            "id": 1,
                                            "name": "Admin",
                                          },
                                          "2": Object {
                                            "id": 2,
                                            "name": "Editor",
                                          },
                                          "3": Object {
                                            "id": 3,
                                            "name": "Viewer",
                                          },
                                        }
              }
              saveSelectData={[Function]}
              setID={1}
              type="roles"
/>,
            <SelectComponent
              IDList={
                            Array [
                                          1,
                                          2,
                                          3,
                                        ]
              }
              data={
                            Object {
                                          "1": Object {
                                            "id": 1,
                                            "name": "Trip to space",
                                          },
                                          "2": Object {
                                            "id": 2,
                                            "name": "Assembly Ikea furniture",
                                          },
                                          "3": Object {
                                            "id": 3,
                                            "name": "Datumize Zentral",
                                          },
                                        }
              }
              saveSelectData={[Function]}
              setID={1}
              type="projects"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "set1-users",
            "nodeType": "class",
            "props": Object {
              "IDList": Array [
                1,
                2,
                3,
              ],
              "data": Object {
                "1": Object {
                  "id": 1,
                  "name": "John Doe",
                },
                "2": Object {
                  "id": 2,
                  "name": "Alice",
                },
                "3": Object {
                  "id": 3,
                  "name": "Bob",
                },
              },
              "saveSelectData": [Function],
              "setID": 1,
              "type": "users",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "set1-roles",
            "nodeType": "class",
            "props": Object {
              "IDList": Array [
                1,
                2,
                3,
              ],
              "data": Object {
                "1": Object {
                  "id": 1,
                  "name": "Admin",
                },
                "2": Object {
                  "id": 2,
                  "name": "Editor",
                },
                "3": Object {
                  "id": 3,
                  "name": "Viewer",
                },
              },
              "saveSelectData": [Function],
              "setID": 1,
              "type": "roles",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "set1-projects",
            "nodeType": "class",
            "props": Object {
              "IDList": Array [
                1,
                2,
                3,
              ],
              "data": Object {
                "1": Object {
                  "id": 1,
                  "name": "Trip to space",
                },
                "2": Object {
                  "id": 2,
                  "name": "Assembly Ikea furniture",
                },
                "3": Object {
                  "id": 3,
                  "name": "Datumize Zentral",
                },
              },
              "saveSelectData": [Function],
              "setID": 1,
              "type": "projects",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <styled.h2>
            Set Number 
            1
</styled.h2>,
          <styled.div>
            <SelectComponent
                        IDList={
                                    Array [
                                                1,
                                                2,
                                                3,
                                              ]
                        }
                        data={
                                    Object {
                                                "1": Object {
                                                  "id": 1,
                                                  "name": "John Doe",
                                                },
                                                "2": Object {
                                                  "id": 2,
                                                  "name": "Alice",
                                                },
                                                "3": Object {
                                                  "id": 3,
                                                  "name": "Bob",
                                                },
                                              }
                        }
                        saveSelectData={[Function]}
                        setID={1}
                        type="users"
            />
            <SelectComponent
                        IDList={
                                    Array [
                                                1,
                                                2,
                                                3,
                                              ]
                        }
                        data={
                                    Object {
                                                "1": Object {
                                                  "id": 1,
                                                  "name": "Admin",
                                                },
                                                "2": Object {
                                                  "id": 2,
                                                  "name": "Editor",
                                                },
                                                "3": Object {
                                                  "id": 3,
                                                  "name": "Viewer",
                                                },
                                              }
                        }
                        saveSelectData={[Function]}
                        setID={1}
                        type="roles"
            />
            <SelectComponent
                        IDList={
                                    Array [
                                                1,
                                                2,
                                                3,
                                              ]
                        }
                        data={
                                    Object {
                                                "1": Object {
                                                  "id": 1,
                                                  "name": "Trip to space",
                                                },
                                                "2": Object {
                                                  "id": 2,
                                                  "name": "Assembly Ikea furniture",
                                                },
                                                "3": Object {
                                                  "id": 3,
                                                  "name": "Datumize Zentral",
                                                },
                                              }
                        }
                        saveSelectData={[Function]}
                        setID={1}
                        type="projects"
            />
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Set Number ",
              1,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Set Number ",
            1,
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <SelectComponent
                IDList={
                                Array [
                                                1,
                                                2,
                                                3,
                                              ]
                }
                data={
                                Object {
                                                "1": Object {
                                                  "id": 1,
                                                  "name": "John Doe",
                                                },
                                                "2": Object {
                                                  "id": 2,
                                                  "name": "Alice",
                                                },
                                                "3": Object {
                                                  "id": 3,
                                                  "name": "Bob",
                                                },
                                              }
                }
                saveSelectData={[Function]}
                setID={1}
                type="users"
/>,
              <SelectComponent
                IDList={
                                Array [
                                                1,
                                                2,
                                                3,
                                              ]
                }
                data={
                                Object {
                                                "1": Object {
                                                  "id": 1,
                                                  "name": "Admin",
                                                },
                                                "2": Object {
                                                  "id": 2,
                                                  "name": "Editor",
                                                },
                                                "3": Object {
                                                  "id": 3,
                                                  "name": "Viewer",
                                                },
                                              }
                }
                saveSelectData={[Function]}
                setID={1}
                type="roles"
/>,
              <SelectComponent
                IDList={
                                Array [
                                                1,
                                                2,
                                                3,
                                              ]
                }
                data={
                                Object {
                                                "1": Object {
                                                  "id": 1,
                                                  "name": "Trip to space",
                                                },
                                                "2": Object {
                                                  "id": 2,
                                                  "name": "Assembly Ikea furniture",
                                                },
                                                "3": Object {
                                                  "id": 3,
                                                  "name": "Datumize Zentral",
                                                },
                                              }
                }
                saveSelectData={[Function]}
                setID={1}
                type="projects"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "set1-users",
              "nodeType": "class",
              "props": Object {
                "IDList": Array [
                  1,
                  2,
                  3,
                ],
                "data": Object {
                  "1": Object {
                    "id": 1,
                    "name": "John Doe",
                  },
                  "2": Object {
                    "id": 2,
                    "name": "Alice",
                  },
                  "3": Object {
                    "id": 3,
                    "name": "Bob",
                  },
                },
                "saveSelectData": [Function],
                "setID": 1,
                "type": "users",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "set1-roles",
              "nodeType": "class",
              "props": Object {
                "IDList": Array [
                  1,
                  2,
                  3,
                ],
                "data": Object {
                  "1": Object {
                    "id": 1,
                    "name": "Admin",
                  },
                  "2": Object {
                    "id": 2,
                    "name": "Editor",
                  },
                  "3": Object {
                    "id": 3,
                    "name": "Viewer",
                  },
                },
                "saveSelectData": [Function],
                "setID": 1,
                "type": "roles",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "set1-projects",
              "nodeType": "class",
              "props": Object {
                "IDList": Array [
                  1,
                  2,
                  3,
                ],
                "data": Object {
                  "1": Object {
                    "id": 1,
                    "name": "Trip to space",
                  },
                  "2": Object {
                    "id": 2,
                    "name": "Assembly Ikea furniture",
                  },
                  "3": Object {
                    "id": 3,
                    "name": "Datumize Zentral",
                  },
                },
                "saveSelectData": [Function],
                "setID": 1,
                "type": "projects",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
